jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java

```diff
@@ -119,77 +193,68 @@ public boolean isPassivating()
+
+        expired.addAll( candidates.stream().filter( candidate -> {
             
             if (LOG.isDebugEnabled())
+            {
                 LOG.debug( "Checking expiry for candidate {}", candidate );
-            
+            }
+            if(expired.contains( candidate ))
+            {
+                return false;
+            }
             try
             {
-                SessionData sd = load(candidate);
-
+                boolean hasCandidate = sessionDataMap.containsKey( getCacheKey( candidate ) );
                 //if the session no longer exists
-                if (sd == null)
+                if(!hasCandidate)
                 {
                     if (LOG.isDebugEnabled())
                     {
                         LOG.debug( "Session {} does not exist in Hazelcast", candidate );
                     }
                     return true;
                 }
-                else
-                {
-                    if (_context.getWorkerName().equals(sd.getLastNode()))
-                    {
-                        //we are its manager, add it to the expired set if it is expired now
-                        if ((sd.getExpiry() > 0 ) && sd.getExpiry() <= now)
-                        {
-                            if (LOG.isDebugEnabled())
-                            {
-                                LOG.debug( "Session {} managed by {} is expired", candidate, _context.getWorkerName() );
-                            }
-                            return true;
-                        }
-                    }
-                    else
-                    {
-                        //if we are not the session's manager, only expire it iff:
-                        // this is our first expiryCheck and the session expired a long time ago
-                        //or
-                        //the session expired at least one graceperiod ago
-                        if (_lastExpiryCheckTime <=0)
-                        {
-                            if ((sd.getExpiry() > 0 ) && sd.getExpiry() < (now - (1000L * (3 * _gracePeriodSec))))
-                            {
-                                return true;
-                            }
-                        }
-                        else
-                        {
-                            if ((sd.getExpiry() > 0 ) && sd.getExpiry() < (now - (1000L * _gracePeriodSec)))
-                            {
-                                return true;
-                            }
-                        }
-                    }
-                }
             }
             catch (Exception e)
             {
                 LOG.warn("Error checking if candidate {} is expired so expire it", candidate, e);
                 return true;
             }
             return false;
-        } ).collect( Collectors.toSet() );
+        } ).collect(Collectors.toSet()));
+
+        if(predicate != null)
```

Comment:
I don't think this is exactly the way I'd like to see this go.  Ideally you would make a method on HazelcastSessionDataStore called something like findExpiredSessionsAt (long time). In that method you can use a predicate to look for sessions that expired before the given time.   For now, you can invoke that method in doGetExpired after you've checked each of the session candidates. The time that you would pass in would be calculated as you've got around line 92.

Eventually, before we release this code, I will push the findExpiredSessionsAt(long time) method up to the AbstractSessionDataStore and implement it for all of the other SessionDataStores. I will also move the calculation of the time up there, as well as the call to it.

Reply:
I'm a bit lost now :-) 
I was thinking about the predicate as a sort of pluggable QueryManager. So that's optional.
For clarity I have removed this "default" predicate from HazelcastSessionDataStore to the test class where it's used.
I don't understand because In the code it's already called in doGetExpired after check of session candidates?